### DISCUSSION TITLE
Application render framerates seem to depend on element spacing within a panel

### DISCUSSION THREAD

**Discussion by mw1920** [AvaloniaUI/Avalonia#13125] [General]
I am writing an Avalonia app that needs to render bitmap-based animations and custom draw operations at a faster frame rate, typically 30-60 FPS. I have noticed that the final frame rates delivered don't necessarily correlate with just the number of running animations, but also with the amount of space around the elements. I have attached a sample to demonstrate this (which only includes a bitmap based widget). On my reasonably fast machine at standard HD resolution, the frame rate is what I expect, but at 4K and full screen, the frame rate drops, but can be restored by resizing the window to eliminate the space around the animations. The problem gets significantly worse on lower power hardware (e.g. RPi 4) where the frame rate varies from expected to very slow as the window is resized larger. 
I cannot easily explain why this is the case with my limited knowledge of Avalonia, so I am presenting the problem here in the hope that someone could  throw light on this, or perhaps tell me raise an issue.

[AvaloniaTestApp.zip](https://github.com/AvaloniaUI/Avalonia/files/12794672/AvaloniaTestApp.zip)

**Comment by timunie** [comment]
A 4K screen is huge task for a low end device. The number of pixels is square, so double screen size means 4 times more pixels to calculate.

**Reply by mw1920** [reply]
Yes, I understand. I didn't try at 4K on RPi 4, only HD - should have made this clearer. RPi 4 is fine and sees 60FPS when there's no intervening space between the animation controls, but slows down considerably as the window is sized larger and more and more of the Panel that contains them is invalidated. It looks like the achievable frame rate is proportional to the actual size of the Window, as well as the number of the animations being run? The point I'm trying to make is what I'm seeing is just what happens as the render code has to deal with more pixels? i.e. there's no bug here?

**Reply by timunie** [reply]
I haven't checked your sample. Also I have no Raspberry to try, so can't finally answer. I can however suggest to take a look into src/samples/RenderDemo and ControlCatalog (animations page)

**Reply by mw1920** [reply]
I'll do some more tests with the samples you mentioned and I will see whether they behave differently.

**Comment by kekekeks** [comment]
Avalonia attempts to only draw the changed area of the screen. However it only tracks ONE area per frame. When you have two changed elements at the opposite corners of the screen it will trigger a full repaint.

We used to have accurate dirty rect tracking, but it involved its own problem with repainting complex backgrounds shared by multiple elements.

Verify that you are using GPU-powered rendering, it should help with the frame rate.

**Reply by mw1920** [reply]
That does make sense now and correlates with what I see in frame rate. 
I guess that if I need to get around this, then I would need to have my animating visuals done another way, such as within a NativeControlHost instance, which I think would not be affected by the rest of the window contents in the same way as a regular Visual? 

I double-checked for GPU rendering being used. It's being used on both Windows and RPi 4 platform. 

As a side point, I found that the best FPS performance on both Windows and Linux for this particular application was to use the non-default renderer settings, namely Win32RenderingMode.Wgl and X11RenderingMode.Egl (but the FPS is still not enough for my needs).

**Reply by maxkatz6** [reply]
WGL on windows would limit you with 60 fps max IIRC. Because composition renderer won't be used.

**Comment by mw1920** [comment]
> WGL on windows would limit you with 60 fps max IIRC. Because composition renderer won't be used.

Interesting, I was under the impression that the rendering is limited by the render loop timer to 60 fps in all cases. So Angle has a higher tick rate, or a way of rendering something outside the render loop?

**Reply by kekekeks** [reply]
We are using DWM-based timer when rendering goes through DWM composition APIs. Otherwise fallback timer is used.

**Reply by mw1920** [reply]
OK many thanks for explaining.

**Comment by tster123** [comment]
To save any future person's time who stumbles upon this discussion, I struggled with a similar issue for many hours and ended up asking a question [here](https://github.com/AvaloniaUI/Avalonia/discussions/17808) and was told how to re-enable the multi-box clipping.  There is an option when you build the app:

```C#
public static AppBuilder BuildAvaloniaApp()
    => AppBuilder.Configure<App>()
        .UsePlatformDetect()
        .With(new CompositionOptions()
        {
            UseRegionDirtyRectClipping = true
        })
        .UseSkia()
```
